Add new Audio Tokenziers

poonehmousavi opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

-->

Fixes #<issue_number>

<!-- Does your PR introduce any breaking changes? If yes, please list them. -->

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

<!--

🎩 Magic happens when you code. Keep the spells flowing!

-->

poonehmousavi commented:
@mravanelli  and @TParcollet, this PR is ready for review. It would be super helpful to review it ASAP since many tasks depend on these new features.

TParcollet commented:
I am on it, but this will take some time because it's a lot of new code.

TParcollet commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+"""This lobe enables the integration of  pretrained WavTokenizer.
```
FileName is wrong.

poonehmousavi replied:
What is the wrong with the filename?

TParcollet replied:
wavTokenzier.py --> wavTokenizer . Also @pplantinga maybe we want to have a look at unifying the naming in lobe as part of the integration folder? Like camelcase for all models?

TParcollet commented on speechbrain/lobes/models/huggingface_transformers/mimi.py:
```diff
@@ -0,0 +1,174 @@
+"""This lobe enables the integration of huggingface pretrained Mimi.
+
+Mimi codec is a state-of-the-art audio neural codec, developed by Kyutai, that combines semantic and acoustic information into audio tokens running at 12Hz and a bitrate of 1.1kbps.
```
split the line.

poonehmousavi replied:
fixed

TParcollet commented on speechbrain/lobes/models/huggingface_transformers/mimi.py:
```diff
@@ -0,0 +1,174 @@
+
+
+class Mimi(HFTransformersInterface):
+    """An wrapper for the HuggingFace Mimi model
```
An -> A. More documentation is necessary. This docstring is what will be visible in the doc when searching for this module so it's more important than the top file description even.

TParcollet commented on speechbrain/lobes/models/huggingface_transformers/mimi.py:
```diff
@@ -0,0 +1,174 @@
+        The audio sampling rate
+    freeze : bool
+        whether the model will be frozen (e.g. not trainable if used
+        as part of training another model)
```
double space instead of 1

poonehmousavi replied:
fixed

TParcollet commented on speechbrain/lobes/models/huggingface_transformers/mimi.py:
```diff
@@ -0,0 +1,174 @@
+        whether the model will be frozen (e.g. not trainable if used
+        as part of training another model)
+    num_codebooks : int  (default: 8)
+        Number of qunatizer. It could be [2,3,4,5,6,7,8]
```
Typo @pplantinga test failing to catch it? :'(

poonehmousavi replied:
what is the typo here? double space?

TParcollet replied:
Dunno what a qunatizer is.

pplantinga replied:
I guess the spell checker doesn't know how to spell `quantizer`

TParcollet commented on speechbrain/lobes/models/huggingface_transformers/mimi.py:
```diff
@@ -0,0 +1,174 @@
+        Returns
+        -------
+        tokens : torch.Tensor
+            A (Batch x NQ x Length) tensor of audio tokens
```
NQ is unclear;

poonehmousavi replied:
replaced with num_codebooks

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+from torch.nn.utils import remove_weight_norm, weight_norm
+
+
+def download_and_extract(repo_id, filename, save_path):
```
This must be refactored to use the SpeechBrain fetch function. If the fetch does not cover this use case, we should make it possible. No utility functions in lobes unless it is impossible to do otherwise.

poonehmousavi replied:
I could use SB fetch thc function instead of hf_hub_download, however they upload zip file and after downloading it needs to be extracted... one option is to skip downloading and ask users to download and extract the ckp and config file

poonehmousavi replied:
I replaced hf_hub_downloa with SB fetch but still the problem of extracting exists.

TParcollet replied:
@poonehmousavi this is still pending?

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+    print(f"File downloaded, extracted to '{save_path}', and ZIP file removed.")
+
+
+def decimal_to_ternary_matrix(decimals, D):
```
You may want to put these functions after the main class for clarity reasons.

poonehmousavi replied:
done

TParcollet replied:
@poonehmousavi I still don't see it, even though it is marked as 'done'?

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+    return int((kernel_size * dilation - dilation) / 2)
+
+
+class round_func5(InplaceFunction):
```
Couldn't you just use "torch.round(5 * input) / 5" in the code when necessary, this is building an overcomplex torch custom operation for something that could be automatically inferred from torch.

TParcollet replied:
If these function ends up being mandatory, unittests will certainly be necessary to check the gradients and shape etc.

poonehmousavi replied:
When you use torch.round(5 * input) / 5 directly, the gradient for torch.round is not defined (or zero). PyTorch treats it as a non-differentiable operation, and gradients will not propagate through it which is essential in scalar quantization.

poonehmousavi replied:
i created a "CustomRoundingFunction" that handle all these rounding and backprop

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return grad_input
+
+
+class round_func9(InplaceFunction):
```
Same as above.

poonehmousavi replied:
same thing here

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return grad_input
+
+
+class round_func_binary(InplaceFunction):
```
Same as above.

poonehmousavi replied:
same

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return grad_input
+
+
+class Heaviside(InplaceFunction):
```
same.

poonehmousavi replied:
same

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+
+
+class PreProcessor(nn.Module):
+    """
```
Maybe the docstring should integrate some contextualisation information about where this is used in the model.

poonehmousavi replied:
added

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+
+class PostProcessor(nn.Module):
+    """
+    A module for postprocessing data through convolution and reshaping.
```
Maybe the docstring should integrate some contextualisation information about where this is used in the model.

poonehmousavi replied:
added

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return output
+
+
+class DownsampleLayer(nn.Module):
```
Use https://github.com/speechbrain/speechbrain/blob/6a3fbba1528eb5d7c3b60921223748954c10a550/speechbrain/lobes/downsampling.py#L15 or add a custom element in there.

poonehmousavi replied:
same as Conv1d

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+            remove_weight_norm(self.layer)
+
+
+class UpsampleLayer(nn.Module):
```
Same as above.

poonehmousavi replied:
same as conv1d

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+
+class ResidualUnit(nn.Module):
+    """
+    A residual unit with two convolutional layers and activation functions.
```
More context on where this is usefull in the model.?

poonehmousavi replied:
added

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return x
+
+
+class Conv1d(nn.Conv1d):
```
SpeechBrain as a Conv1D nnet already, this may be highly confusing to users. Why can't we use the already existing one?

poonehmousavi replied:
I've transferred their original code to SpeechBrain. I'll check to see if this integration is feasible.

poonehmousavi replied:
If I want to use SB, conv1, I need to add init with zaviaer_uniform to conv1d, is that ok?

poonehmousavi replied:
I have tried it, the mechanism is the same but if we want to load the pre-trained weights, it raises issues due to different naming... so if we want to train it from scratch, it is ok to use our own implementation, but when we want to load the checkpoints, we need to follow their code.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return super(Conv1d, self).forward(x)
+
+
+class ConvTranspose1d(nn.ConvTranspose1d):
```
Same than above.

poonehmousavi replied:
same as Conv1d

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return x
+
+
+class ScalarModel(nn.Module):
```
The docstring needs an example. This is true for a few other classes in this file.

poonehmousavi replied:
added for ScalarModel.. what are the other modules that need docstring? Or in general, what is the rule? I have added docstring for main modules that might be used as an interface.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+    return ternary_matrix
+
+
+def ternary_matrix_to_decimal(matrix):
```
You will want to develop unittests for this function and put them in the test suit.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+    print(f"File downloaded, extracted to '{save_path}', and ZIP file removed.")
+
+
+def decimal_to_ternary_matrix(decimals, D):
```
You will want to develop unittests for this function and put them in the test suit.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        return x
+
+
+class SQCodec(nn.Module):
```
This class should be at the top.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+    save_path : str, optional
+        Directory where the model and configuration files are saved (default is None).
+    config : str, optional
+        Configuration filename for the model (default is 'config.yaml').
```
Config of what? HF? SB?

TParcollet replied:
Is this standard to HF or is it a requirement of our implementation? Anyway, this needs to be explicated in the doctstring, maybe giving an example of a path in a HF repo?

poonehmousavi replied:
In general, their code is not very well organized and they are different version of that. The version that  I used is the code that sent to me by the author. They upload everything on HF without providing any interface and in their code, they expect that you download everything manually.

poonehmousavi replied:
I added a little bit more info in the docstring.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        ScalarModel
+            The built scalar codec model loaded with weights from the checkpoint.
+        """
+        exp_model_config = OmegaConf.load(config)
```
Hum, this is quite borring to introduce yet another way of loading config for a single model. Can't this be managed via HuggingFace instead?

poonehmousavi replied:
it is necessary for loading the their checkpoints

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        self.dim_codebook = dim_codebook
+        self.n_codebook = n_codebook
+        self.bw = bw
+        self.freq = self.n_codebook * 50
```
This 50 seems to be arbitrary? Shouldn't it be a parameter? Nothing must be coded in hard.

poonehmousavi replied:
there is no usecase for this attribute, so I remove it

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        """
+        exp_model_config = OmegaConf.load(config)
+        scalar_codec = ScalarModel(**exp_model_config.generator.config)
+        parameter_dict = torch.load(self.ckpt_path)
```
This operation looks like something that our Pretrainer class should be able to handle, it would be wiser to use it for clarity and our users habits.

TParcollet commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        out = self.scalar_codec.decode(emb_quant.float().to(self.device))
+        return out.detach().cpu().squeeze(0)
+
+    def infer(self, wav_root):
```
Infer is not very self descriptive, especially since there is also an encode function. Maybe rename to same clearer?

poonehmousavi replied:
changed to "reconstruct"

TParcollet reviewed (changes requested):
Thanks for the work again. Quite a few minor things. As mentioned already, I don't think that these new lobes should be put in a discrete folder. They should either go to the HF folder if using transformer or to the general lobe folder. This raise a good question around how to start regrouping models under a better directory tree, but we must all talk about that -- and I am pretty sure that the output type shouldn't be a criterion.

poonehmousavi commented on speechbrain/lobes/models/discrete/speechTokenizer.py:
```diff
@@ -61,6 +58,18 @@ def __init__(
     ):
+
+        # Lazy import to avoid circular dependency issues
+        try:
```
same as the above comments.
Regarding implementation on SpeechBrain, it should be feasible, but it would require porting a significant amount of code from the original repository. We need to establish a clear policy concerning audio tokenizers. This is a rapidly evolving field, with new tokenizers introduced almost monthly. While implementing all from scratch could make our code cleaner and less dependent, it might hinder our ability to keep pace with the field. Moreover, many of them introducing a new version, we need to make sure we are updated with their changes.

TParcollet replied:
Well the entire field is evolving pretty rapidly, which is not preventing us from implementing things in SB. For SSL models, we got lucky enough that HuggingFace decided to do the transformers library. I'll write a comment on the Slack about an idea I just got for this problem.

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+        The name of the config file.
+    checkpoint : str
+        The name of the checkpoint file.
+    sample_rate : int (default: 24000)
```
The general trend is to use 16kHz, especially in the speech domain. However, it's common for some tokenizers to operate at higher frequencies like 24kHz or even 44kHz, such as DAC, particularly if they are used in other domains like music.

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+        path = snapshot_download(repo_id=source, cache_dir=save_path)
+        checkpoint_path = os.path.join(path, checkpoint)
+        config_path = os.path.join(path, config)
+        self.model = self.wavtokenizer.WavTokenizer.from_pretrained0802(
```
Please note that the functions for loading the model are specific to their implementation and are not part of the Hugging Face Transformers library.

TParcollet replied:
Can't we rely on the official one? Checking what they do that is so special to re do it afterwards. Asking this because relying on obscure external function implies that we will 100% have to fix bugs in the near future because they changed their API or because it does not work anymore with the format of the file in the HF hub etc etc. Like it will 100000% happen. If the code is in speechbrain, we can maintain it.

poonehmousavi commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+        config="config.yaml",
+        checkpoint="ckpt_00190000.pth",
+        sample_rate=16000,
+        dim_codebook=19683,
```
It is based on their original setting,

poonehmousavi replied:
It is based on the scalar quantization space

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+"""This lobe enables the integration of  pretrained WavTokenizer.
+
+Note that you need topip install git+https://github.com/Tomiinek/WavTokenizer` to use this module.
```
fixed

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+
+
+class WavTokenizer(nn.Module):
+    """An wrapper for the WavTokenizer model
```
fixed

pplantinga replied:
This is a grammar error, not spelling

TParcollet replied:
Don't see it fixed.

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+
+
+class WavTokenizer(nn.Module):
+    """An wrapper for the WavTokenizer model
```
added more info

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,157 @@
+        self,
+        source,
+        save_path=None,
+        config="wavtokenizer_smalldata_frame40_3s_nq1_code4096_dim512_kmeans200_attn.yaml",
```
It is the smallest one

poonehmousavi commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+"""This lobe enables the integration of speech codec model (SQ-Codec) with scalar quantization,.
```
changed to sq_codec.py

poonehmousavi commented on speechbrain/lobes/models/discrete/SQ-Codec.py:
```diff
@@ -0,0 +1,1497 @@
+
+class SQCodec(nn.Module):
+    """
+    Speech codec model (SQ-Codec) with scalar quantization. It maps the complex speech signal into a finite and compact latent space.
```
added

poonehmousavi commented:
@TParcollet, I have implemented most of the changes you requested. Please double-check and let me know if any further modifications are needed. Here are some general points that were repeated in the reviews, which I would like to address here:

- Regarding the use of the SB module instead of the original code modules like Conv1d, etc.: It is feasible to switch to SB modules if we plan to retrain the model, ensuring that the checkpoints are consistent with the new naming convention. However, as we are currently using existing checkpoints, replacing the modules introduces naming conflicts and issues with loading the checkpoints properly. While there might be a workaround, I'm not sure how much effort we want to invest in this since all of these modules will eventually be transferred to the integration folder. Therefore, I prefer to maintain the original model structure. If we decide to reimplement their code in SB, we could then consider replacing the modules with the corresponding SB modules.

- Many of the tokenizers currently lack a Hugging Face Transformer interface. They only provide their weights and some code to load them, which does not follow the usual loading routine. I assume that HuggingFace will soon offer some interface at least for some of them.

TParcollet reviewed (changes requested):
@mravanelli requested that I do another review of this PR. I started, but the problem is that I don't see the fixes to my previous comments (certainly not pushed yet as commented as done?). I therefore stopped doing the review as to limit the number of review rounds. Please @poonehmousavi let me know when it's pushed so I can do a last review.

Thanks!

TParcollet commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,166 @@
+
+class WavTokenizer(nn.Module):
+    # """A wrapper for the WavTokenizer model
+    """This lobe enables the integration of pretrained WavTokenizer model, a discrete codec models with single codebook for Audio Language Modeling.
```
Adapt the doctstring to integrations already?

poonehmousavi replied:
@pplantinga  is there any specific format i need to follow here?

pplantinga replied:
I think the only change I would suggest is removing "lobe" from the docstring, instead we could call it a "module" or "class"

TParcollet reviewed (changes requested):
Hi @poonehmousavi I did another pass, thanks for this import work. I think most comments are minor, but at least one is quite important and will require a bit of re-thinking. Once my comments are addressed, I think we can merge it. But please do ping me, I'll make a final review :-)

poonehmousavi commented:
@pplantinga I'm adding new tokenizers in this PR, and I believe all of them need to be transferred to the integration folder. I’ve added you to this PR to ensure that we’re aligned on the settings and the steps required for the transfer. This will help us converge the effort efficiently.

pplantinga commented:
> @pplantinga I'm adding new tokenizers in this PR, and I believe all of them need to be transferred to the integration folder. I’ve added you to this PR to ensure that we’re aligned on the settings and the steps required for the transfer. This will help us converge the effort efficiently.

From the perspective of the integrations folder I have no objections to this PR moving forward as-is. Once its merged they can be moved to the integrations folder without too much trouble.

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        emb, emb_quant, x = self.scalar_codec.inference(wav)
+        return x.detach().cpu().squeeze(0)
+
+    @property
```
it will return the quantized singal (x)

TParcollet replied:
This function does nothing in that case as it's not even based on a class attribute that may change. I propose to just remove this function. (it's always true, so no point having it).

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+
+class ScalarModel(nn.Module):
+    """
+    A custom neural network model for encoding and decoding audio signals.
```
links to paper and github are mentioned here:

- https://github.com/poonehmousavi/speechbrain/blob/71ac4a5ea60978fbd9d8980486a4a0dca35cd8fe/speechbrain/lobes/models/discrete/sq_codec.py#L6
- https://github.com/poonehmousavi/speechbrain/blob/71ac4a5ea60978fbd9d8980486a4a0dca35cd8fe/speechbrain/lobes/models/discrete/sq_codec.py#L34

TParcollet replied:
Sure, but a SpeechBrain user will not look at another class docstring when looking at this one. Can we add the relevant reference to this class docstring as well?

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        return x
+
+
+class CustomRoundingFunction(Function):
```
Since it only used for this module so far, lets keep it here. I am working on another recepie that include many of these quantization techniques.. i will take care of this in that PR.

TParcollet replied:
I'll trust you on this. It is critical, don't forget the unit test as it definitely is a function that could go yolo in the future. Please ping me on the upcoming PR so I can double check as well.

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        return output
+
+
+class DownsampleLayer(nn.Module):
```
If I use the SB downsampling, it causes an error when loading the pretrained weights since the naming could be a bit different.  Since it is going to be moved to the integration folder, I suggest keep all their module as it is . 
I am working on another recepie that include many of these quantization techniques.. i will take care of this in that PR.

TParcollet replied:
Well @pplantinga is *just* supposed to transpose code to another folder (with the integration folder), not to re-develop some modules -- especially ones that he does not master. I'll trust you again on that but unit tests and this are important issues, if we don't see a PR addressing these issues at some point we will have to revert. It also becomes harder to track for the reviewers because your new PR will certainly add many new features that will require review, and we will forget about this -- so please, when creating the PR, refer to this one and this comment.

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+            remove_weight_norm(self.layer)
+
+
+class UpsampleLayer(nn.Module):
```
If I use the SB downsampling, it causes an error when loading the pretrained weights since the naming could be a bit different.  Since it is going to be moved to the integration folder, I suggest keep all their module as it is . 
I am working on another recepie that include many of these quantizati

TParcollet replied:
**Tag for later -- this must be verified in new PR.**

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        return x
+
+
+class Conv1d(nn.Conv1d):
```
If I use the SB downsampling, it causes an error when loading the pretrained weights since the naming could be a bit different.  Since it is going to be moved to the integration folder, I suggest keep all their module as it is . 
I am working on another recepie that include many of these quantizati

TParcollet replied:
That is starting to be a lot of things. If not done in next PR, we will have to revert.

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        return super(Conv1d, self).forward(x)
+
+
+class ConvTranspose1d(nn.ConvTranspose1d):
```
If I use the SB downsampling, it causes an error when loading the pretrained weights since the naming could be a bit different.  Since it is going to be moved to the integration folder, I suggest keep all their module as it is . 
I am working on another recepie that include many of these quantization techniques.. i will take care of this in that PR.

poonehmousavi replied:
same as prev one.

TParcollet replied:
**tag for further review next PR**

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        return x
+
+
+def download_and_extract(repo_id, filename, save_path):
```
Make sense.. i will go with second option.

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+
+def download_and_extract(repo_id, filename, save_path):
+    """
+    Downloads a ZIP file from the specified repository, extracts its contents,P
```
removed

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+        filename,
+        save_path=None,
+        config="config.yaml",
+        checkpoint="ckpt_00190000.pth",
```
done

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+"""This lobe enables the integration of speech codec model (SQ-Codec) with scalar quantization,.
```
@pplantinga  is there any specific format i need to follow here?

poonehmousavi commented on conftest.py:
```diff
@@ -62,5 +62,14 @@ def pytest_generate_tests(metafunc):
     )
+try:
+    import wavtokenizer  # noqa: F401
+except ModuleNotFoundError:
```
@pplantinga  is there any specific format i need to follow here?

pplantinga replied:
This file will have significant changes anyway in the integrations PR, so I'll make the required changes in the integrations folder PR once merged.

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+    Arguments
+    ---------
+    repo_id : str
+        The repository ID from which to download the ZIP file.
```
removed it

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+    print(f"File downloaded, extracted to '{save_path}', and ZIP file removed.")
+
+
+def decimal_to_ternary_matrix(decimals, D):
```
nce this module is planned to be moved to the integration folder, I recommend keeping all related modules unchanged for now.

I’m currently working on another recipe that integrates many of these quantization techniques. In that upcoming PR, I’ll address all these points comprehensively, including unittests, to ensure everything works seamlessly.

TParcollet replied:
**tag for newer PR**

poonehmousavi commented on speechbrain/lobes/models/discrete/sq_codec.py:
```diff
@@ -0,0 +1,1394 @@
+    return ternary_matrix
+
+
+def ternary_matrix_to_decimal(matrix):
```
same

TParcollet replied:
**tag for newer PR**

poonehmousavi commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,166 @@
+
+
+class WavTokenizer(nn.Module):
+    # """A wrapper for the WavTokenizer model
```
done

poonehmousavi commented on speechbrain/lobes/models/huggingface_transformers/mimi.py:
```diff
@@ -0,0 +1,186 @@
+
+
+class Mimi(HFTransformersInterface):
+    # """An wrapper for the HuggingFace Mimi model
```
done

poonehmousavi commented:
@TParcollet 

I have addressed most of your points. Regarding the unittests and the suggestion to use the SB module instead of the provided module, as I mentioned in the comments, I’m currently working on another recipe that incorporates several quantization techniques. This requires re-implementing many of these techniques in SB from scratch. 

Once that work is complete, I’ll focus on modifying these models accordingly. For now, since these models are in the integration folder, there shouldn’t be any issues.

poonehmousavi commented on speechbrain/lobes/models/discrete/speechTokenizer.py:
```diff
@@ -47,12 +44,12 @@ class SpeechTokenizer_interface(nn.Module):
     torch.Size([8, 10, 2])
-    >>> wav=model.decode(tokens)
-    >>> print(wav.shape)
+    >>> wav=model.decode(tokens)  # doctest: +SKIP
```
removed

TParcollet reviewed (changes requested):
Ignore this comment and review, GitHub completely bugged and I couldn't see the updated version of the code. I'll do another review with the correct code.

TParcollet commented on speechbrain/lobes/models/discrete/wavTokenzier.py:
```diff
@@ -0,0 +1,165 @@
+"""This lobe enables the integration of pretrained WavTokenizer.
```
filename (wavTokenzier -> wavTokenizer)

TParcollet reviewed (approved):
@poonehmousavi thanks for your work! This is a 'condition' approval as a few important comments have been pushed to another PR. If this other PR does not come in, we will not be able to release this PR into the master branch (i.e. will be reverted).

poonehmousavi commented:
@TParcollet Since all the remaining points are related to SQ-Codec, I could remove the file and we only merge Wavtokenzier and Mimi for now.. later I will add a new PR for SQ-Codec,

TParcollet commented:
@poonehmousavi sounds good to me!

poonehmousavi commented:
@TParcollet and @mravanelli  I have removed teh sq_codec for now. The other tokenizers I think are ready to merge.

poonehmousavi commented:
No idea why I get an error for MERT here, i didn't change anything just removed a file... is there any new update to CI that could cause this?

pplantinga commented:
> No idea why I get an error for MERT here, i didn't change anything just removed a file... is there any new update to CI that could cause this?

Probably an error on the huggingface side. I have run the test locally on this branch and it passes. I restarted the test, hopefully it passes this time. And in the future, this will be part of the integrations, so failures here won't cause CI failures.

poonehmousavi commented:
still the same problem. i remembered it has the same issue but then we fixed it by adding
`    WARNING: ...` line in the docstring to catch that error

pplantinga commented:
Aha, this is actually due to new version of transformers. I just upgraded my local version and got the same error

pplantinga commented:
This test will be moved to integrations folder soon, so let's find a simple way to skip it for now

poonehmousavi commented:
could we simply add skip test for now?
